crates/amaru-consensus/src/consensus/store.rs

```diff
@@ -37,8 +39,152 @@ impl Display for StoreError {
+    fn encode<W: cbor::encode::Write>(
+        &self,
+        e: &mut cbor::Encoder<W>,
+        ctx: &mut C,
+    ) -> Result<(), cbor::encode::Error<W::Error>> {
+        e.array(4)?;
+        e.encode_with(self.active, ctx)?;
+        e.encode_with(self.evolving, ctx)?;
+        e.encode_with(self.candidate, ctx)?;
+        e.encode_with(self.previous_epoch_last_header, ctx)?;
+        Ok(())
+    }
+}
+
+impl<'b, C> cbor::decode::Decode<'b, C> for Nonces {
+    fn decode(d: &mut cbor::Decoder<'b>, ctx: &mut C) -> Result<Self, cbor::decode::Error> {
+        d.array()?;
+        Ok(Nonces {
+            active: d.decode_with(ctx)?,
+            evolving: d.decode_with(ctx)?,
+            candidate: d.decode_with(ctx)?,
+            previous_epoch_last_header: d.decode_with(ctx)?,
+        })
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum NoncesError {
+    #[error("cannot find nonces: unknown parent {parent} from header {header}")]
+    UnknownParent { header: Hash<32>, parent: Hash<32> },
+
+    #[error("unknown header: {header}")]
+    UnknownHeader { header: Hash<32> },
+
+    #[error("no parent header for: {header} (where one is clearly expected)")]
+    NoParentHeader { header: Hash<32> },
+
+    #[error("{0}")]
+    StoreError(#[from] StoreError),
+}
+
+impl<H: IsHeader> Praos<H> for dyn ChainStore<H> {
+    type Error = NoncesError;
+
+    fn get_nonce(&self, header: &Hash<32>) -> Option<Nonce> {
+        self.get_nonces(header)
+            .map(|(_epoch, nonces)| nonces.active)
+    }
+
+    fn evolve_nonce(&mut self, header: &H) -> Result<(), Self::Error> {
+        let (epoch, is_within_stability_window) = nonce::randomness_stability_window(header);
+
+        let parent_hash = header.parent().ok_or_else(|| NoncesError::NoParentHeader {
+            header: header.hash(),
+        })?;
+
+        let (parent_epoch, parent) =
+            self.get_nonces(&parent_hash)
+                .ok_or_else(|| NoncesError::UnknownParent {
+                    header: header.hash(),
+                    parent: parent_hash,
+                })?;
+
+        // Compute the new evolving nonce by combining it with the current one and the header's VRF
+        // output.
+        // Specifically, we combine it with `η` (a.k.a eta), which is a blake2b-256 hash of the
+        // tagged leader VRF output after a range extension. The range extension is, yet another
+        // blake2b-256 hash.
+        let evolving = nonce::evolve(header, &parent.evolving);
+
+        self.put_nonces(
+            &header.hash(),
+            epoch,
+            Nonces {
+                evolving,
+
+                // On epoch changes, compute the new active nonce by combining:
+                //   1. the (now stable) candidate; and
+                //   2. the previous epoch's last block's parent header hash.
+                //
+                // If the epoch hasn't changed, then our active nonce is unchanged.
+                active: if epoch > parent_epoch {
```

Comment:
This inline definition is large and complicated enough to warrant a comment: this seems to call for a dedicated function or variable